collect_app/src/test/java/org/odk/collect/android/widgets/TriggerWidgetTest.java

```diff
@@ -1,67 +1,98 @@
 package org.odk.collect.android.widgets;
 
-import androidx.annotation.NonNull;
-
 import android.view.View;
 import android.widget.CheckBox;
 
-import org.javarosa.core.model.data.IAnswerData;
 import org.javarosa.core.model.data.StringData;
+import org.javarosa.form.api.FormEntryPrompt;
 import org.junit.Test;
+import org.junit.runner.RunWith;
 import org.odk.collect.android.formentry.questions.QuestionDetails;
-import org.odk.collect.android.widgets.base.QuestionWidgetTest;
+import org.odk.collect.android.listeners.WidgetValueChangedListener;
+import org.robolectric.RobolectricTestRunner;
 
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.is;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertTrue;
-import static org.mockito.Mockito.when;
-
-/**
- * @author James Knight
- */
-
-public class TriggerWidgetTest extends QuestionWidgetTest<TriggerWidget, StringData> {
-    @NonNull
-    @Override
-    public TriggerWidget createWidget() {
-        return new TriggerWidget(activity, new QuestionDetails(formEntryPrompt, "formAnalyticsID"));
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.verify;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.mockValueChangedListener;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithAnswer;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly;
+import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
+
+@RunWith(RobolectricTestRunner.class)
+public class TriggerWidgetTest {
```

Comment:
I've chosen to not extend the `QuestionWidgetTest`. From looking at a lot of our tests, although this pattern initially makes sense to me, it seems we run into problems with widgets having different behaviours for some of the common methods. I thought trying to build out helpers `QuestionWidgetHelpers` would be a more flexible approach - make the tests easy to write rather than making the tests hard to change.

Additionally it feels like the hierarchical nature of the widgets has only really caused problems (unpicking `StringWidget` common code for example) and I think using test extension here feeds into that problem as well.

Interested to see what others think of this however!

Reply:
On one hand it's too bad that we'll end up with a bunch of trivial duplicated tests. But on the other they're really easy to write and there is value to having each widget considered independently so that they can more easily be refactored to suit their specific needs as you say. We just have to be careful to make sure all the same edge cases are considered.